_posts/2024-09-12-opensearch-project-roadmap-2024-2025.md

```diff
@@ -0,0 +1,187 @@
+
+[**Cluster state management**](https://github.com/opensearch-project/OpenSearch/issues/13257): The cluster manager node manages all admin operations in the cluster. These operations include creating and deleting indexes, updating fields in an existing index, taking snapshots, and adding and removing nodes. The metadata about indexes and other entities—data streams, templates, aliases, snapshots, and custom entities stored by plugins—is stored in a data structure called cluster state. Any change to cluster state is processed by the cluster manager node and persisted to that node’s local disk. Starting with version 2.12, OpenSearch added support for storing the cluster state remotely in order to provide durability guarantees. With the introduction of remote cluster state, replacing all cluster manager nodes will not result in any data loss in remote store clusters. The cluster manager node processes any cluster state updates and then sends the updated state to all the follower nodes in the cluster. As the state and number of follower nodes grow, the overhead on the cluster manager node increases significantly, as the cluster manager is responsible for publishing the updated state to all nodes in the cluster. This impacts the cluster’s stability and availability. To reduce work on Cluster Manager node, we are proposing to use the remote store for cluster state publication. The cluster manager will publish the entire cluster state to the remote store for download by each follower node. The published cluster state will include the ephemeral entities like the[shard routing table](https://github.com/opensearch-project/OpenSearch/issues/14164), which stores the mapping of the shards assigned to each data node in the cluster. The cluster manager node will only communicate that a new state is available and provide the remote location of the new state, instead of publishing the entire cluster state. Publishing the state remotely will reduce memory, CPU, and transport thread overhead on the cluster manager node during cluster state changes. This approach also allows for on-demand download of entities on the data or coordinator nodes instead of requiring all nodes to maintain full cluster state. This will align with our vision of a more cloud native architecture by supporting stateless cluster manager backed by remote store in future. Remote publication will be generally available in OpenSearch 2.17 and is planned to be further enhanced in future releases.
```

Comment:
```suggestion
[**Cluster state management**](https://github.com/opensearch-project/OpenSearch/issues/13257): The cluster manager node manages all admin operations in a cluster. These operations include creating and deleting indexes, updating fields in an existing index, taking snapshots, and adding and removing nodes. The metadata about indexes and other entities—data streams, templates, aliases, snapshots, and custom entities stored by plugins—is stored in a data structure called the cluster state. Any change to the cluster state is processed by the cluster manager node and persisted to that node's local disk. Starting with version 2.12, OpenSearch added support for storing the cluster state remotely in order to provide durability guarantees. With the introduction of a remote cluster state, replacing all cluster manager nodes will not result in any data loss in remote store clusters. The cluster manager node processes any cluster state updates and then sends the updated state to all the follower nodes in the cluster. As the state and number of follower nodes grow, the overhead on the cluster manager node increases significantly because the cluster manager node is responsible for publishing the updated state to every node in the cluster. This impacts the cluster's stability and availability. To reduce strain on the cluster manager node, we are proposing to use the remote store for cluster state publication. The cluster manager node will publish the entire cluster state to the remote store to be downloaded by each follower node. The published cluster state will include ephemeral entities like the [shard routing table](https://github.com/opensearch-project/OpenSearch/issues/14164), which stores the mapping of the shards assigned to each data node in the cluster. The cluster manager node will only communicate that a new state is available and provide the remote location of the new state, instead of publishing the entire cluster state. Publishing the state remotely will reduce memory, CPU, and transport thread overhead on the cluster manager node during cluster state changes. This approach will also allow on-demand downloading of entities on the data or coordinator nodes instead of requiring all nodes to maintain the full cluster state. This will align with our vision of a more cloud-native architecture. Remote publication will be generally available in OpenSearch 2.17 and is planned to be further enhanced in future version releases.
```